aptos-move/move-examples/post_mint_reveal_nft/sources/whitelist.move

```diff
@@ -0,0 +1,145 @@
+    use aptos_framework::timestamp;
+    use aptos_framework::account;
+    use post_mint_reveal_nft::bucket_table::{Self, BucketTable};
+    use std::error;
+    use std::signer;
+    use std::vector;
+
+    friend post_mint_reveal_nft::minting;
+
+    /// WhitelistMintConfig stores information about all stages of whitelist.
+    /// Most whitelists are one-stage, but we allow multiple stages to be added in case there are multiple rounds of whitelists.
+    struct WhitelistMintConfig has key {
+        whitelist_configs: vector<WhitelistStage>,
+    }
+
+    /// WhitelistMintConfigSingleStage stores information about one stage of whitelist.
+    struct WhitelistStage has store {
+        whitelisted_address: BucketTable<address, u64>,
+        whitelist_mint_price: u64,
+        whitelist_minting_start_time: u64,
+        whitelist_minting_end_time: u64,
+    }
+
+    /// The whitelist start time is not strictly smaller than the whitelist end time.
+    const EINVALID_WHITELIST_SETTING: u64 = 1;
+    /// The whitelist stage should be added in order. If the whitelist_stage parameter is not equal to the length of the whitelist_configs vector,
+    /// it means that the whitelist stage is not added in order and we need to abort.
+    const EINVALID_STAGE: u64 = 2;
+    /// The specified address does not exist.
+    const EACCOUNT_DOES_NOT_EXIST: u64 = 3;
+    /// The amount that the user wants to mint exceeds the amount that the user is allowed to mint.
+    const EEXCEEDS_MINT_LIMIT: u64 = 4;
+    /// Cannot add more addresses to the whitelist after this whitelist minting stage already ends.
+    const EINVALID_UPDATE_AFTER_MINTING: u64 = 5;
+    /// The given address is not on the whitelist.
+    const EACCOUNT_NOT_WHITELISTED: u64 = 6;
+
+    #[view]
+    /// Checks if WhitelistMintConfig resource exists.
+    public fun whitelist_config_exists(module_address: address): bool {
+        exists<WhitelistMintConfig>(module_address)
+    }
+
+    #[view]
+    /// Checks if the current time is within a whitelist stage. Return value bool is true if the current time is within a whitelist stage.
+    /// Return value u64 is the index of the whitelist stage that the current time is within.
+    public fun get_whitelist_stage_by_current_time(module_address: address): (bool, u64) acquires WhitelistMintConfig {
+        // If the project doesn't have a whitelist, return false.
+        if (!whitelist_config_exists(module_address)) {
+            return (false, 0)
+        };
+        let whitelist_mint_config = borrow_global<WhitelistMintConfig>(module_address);
+        let now = timestamp::now_seconds();
+
+        let i = 0;
+        while (i < vector::length(&whitelist_mint_config.whitelist_configs)) {
+            let whitelist_stage = vector::borrow(&whitelist_mint_config.whitelist_configs, i);
+            if (whitelist_stage.whitelist_minting_start_time <= now && now < whitelist_stage.whitelist_minting_end_time) {
```

Comment:
Why not now <= whitelist_stage.whitelist_minting_end_time?

Reply:
If you have two consecutive ranges do we want them to be [1,4] [5, 10] or [1, 5) and [5, 11) the half-open intervals have nicer properties that the end of the first range is the start of the next range.

Reply:
yeah agree w/ gerben, we want the start to be inclusive and the end to be non-inclusive

Reply:
Makes sense!